lib/pbench/cli/agent/commands/tools/list.py

```diff
@@ -57,80 +57,68 @@ def execute(self) -> int:
+        toolname = self.context.name
+        found = False
         tool_info = {}
 
-        if not self.context.name:
-            for group in groups:
-                tool_info[group] = {}
-                try:
-                    tg_dir = self.gen_tools_group_dir(group).glob("*/**")
-                except BadToolGroup:
-                    self.logger.error("Bad tool group: %s", group)
-                    return 1
+        for group in groups:
+            tool_info[group] = {}
+            try:
+                tg_dir = self.gen_tools_group_dir(group)
+            except BadToolGroup:
+                self.logger.error("Bad tool group: %s", group)
+                return 1
+
+            for path in tg_dir.iterdir():
+                # skip __trigger__ if present
+                if not path.is_dir():
+                    continue
 
-                for path in tg_dir:
-                    host = path.name
-                    tool_info[group][host] = {}
+                host = path.name
+                tool_info[group][host] = {}
 
+                if toolname:
+                    tool_info[group][host]["label"] = None
+                else:
                     label = path / "__label__"
                     v = label.read_text().rstrip("\n") if label.exists() else None
                     tool_info[group][host]["label"] = v
 
-                    tool_info[group][host]["tools"] = {}
-
-                    for tool in sorted(self.tools(path)):
-                        v = sorted((path / tool).read_text().rstrip("\n").split("\n"))
-                        tool_info[group][host]["tools"][tool] = v if opts else ""
-
-            if tool_info:
-                found = self.print_results(tool_info, self.context.with_option)
-                if not found:
-                    msg = "No tools found"
-                    if self.context.group:
-                        msg += f' in group "{self.context.group[0]}"'
-                    self.logger.warn(msg)
-            else:
-                self.logger.warn("No tool groups found")
-
-            return 0
-
-        else:
-            # List the groups which include this tool
-            tool = self.context.name
-            found = False
-            for group in groups:
-                try:
-                    tg_dir = self.gen_tools_group_dir(group)
-                except BadToolGroup:
-                    self.logger.error("Bad tool group: %s", group)
-                    return 1
-
-                tool_info[group] = {}
-                for path in tg_dir.iterdir():
-                    # skip files like __label__ and __trigger__
-                    if not path.is_dir():
-                        continue
-
-                    host = path.name
-                    tool_info[group][host] = {}
-                    tool_info[group][host]["label"] = None
-                    tool_info[group][host]["tools"] = {}
-
+                toolsdict = tool_info[group][host]["tools"] = {}
```

Comment:
If you happen to come in here again, consider this tweak:
```python
                tool_info[group][host] = {
                    "label": None,
                    "tools": {},
                }

                if not toolname:
                    label = path / "__label__"
                    v = label.read_text().rstrip("\n") if label.exists() else None
                    tool_info[group][host]["label"] = v

                toolsdict = tool_info[group][host]["tools"]
```
This initializes the `tool_info[group][host]` keys in a single assignment, and it gets rid of the double assignment of `toolsdict` (as well as simplifying the assignment of `["label"]`).

And, it looks like it might be feasible to combine this `if` statement with the one at line 88:
```python
                tool_info[group][host] = {
                    "label": None,
                    "tools": {},
                }
                toolsdict = tool_info[group][host]["tools"]

                if toolname:
                    # Check if the tool is in any of the hosts.
                    if toolname in self.tools(path):
                        v = (path / toolname).read_text().rstrip("\n").split("\n")
                        toolsdict[toolname] = sorted(v) if opts else ""
                        found = True
                else:
                    # no tool name was specified
                    for tool in sorted(self.tools(path)):
                        v = (path / tool).read_text().rstrip("\n").split("\n")
                        toolsdict[tool] = sorted(v) if opts else ""

                    label = path / "__label__"
                    v = label.read_text().rstrip("\n") if label.exists() else None
                    tool_info[group][host]["label"] = v
```

Or, we could get rid of the other `if` in a different way:
```python
                tool_info[group][host] = {
                    "label": None,
                    "tools": {},
                }

                if toolname:
                    found = toolname in self.tools(path)
                    toolslist = [toolname] if found else []
                else:
                    toolslist = sorted(self.tools(path))

                    label = path / "__label__"
                    v = label.read_text().rstrip("\n") if label.exists() else None
                    tool_info[group][host]["label"] = v

                toolsdict = tool_info[group][host]["tools"]
                for tool in toolslist:
                    v = (path / tool).read_text().rstrip("\n").split("\n")
                    toolsdict[tool] = sorted(v) if opts else ""
```
That is, if there is a `toolname` we set `found` if it's in the list, and then we set up `toolslist` with the tools to visit, which might be the one tool, or no tools, or all the tools in sorted order; if there's no `toolname`, then we do the "label thing"; then we visit all the tools that we're supposed to visit (which might be none).  Simple!  😁  And, we do it with only one `if` and no unnecessary reads or sorts.  🙂